File: shamelagpt-android/app/src/main/java/com/shamelagpt/android/core/error/AppError.kt
package com.shamelagpt.android.core.error

import android.content.Context
import com.shamelagpt.android.R
import java.io.IOException

/**
 * Standardized error types for the ShamelaGPT application.
 */
sealed class AppError : Exception() {

    /** Debug code used for support tickets and logging */
    abstract val debugCode: String

    /** Localized, user-facing message */
    abstract fun getUserMessage(context: Context): String

    data class Network(val originalError: Throwable? = null) : AppError() {
        override val debugCode: String = "E-APP-001"
        override val message: String? = originalError?.message ?: "Network error occurred"

        override fun getUserMessage(context: Context): String {
            return context.getString(R.string.no_internet)
        }
    }

    data class Auth(val originalError: Throwable? = null) : AppError() {
        override val debugCode: String = "E-APP-002"
        override val message: String? = originalError?.message ?: "Authentication error"

        override fun getUserMessage(context: Context): String {
            return context.getString(R.string.error_generic_message)
        }
    }

    data class Api(val code: Int, val apiMessage: String?) : AppError() {
        override val debugCode: String = "E-APP-$code"
        override val message: String? = apiMessage ?: "API error ($code)"

        override fun getUserMessage(context: Context): String {
            return apiMessage ?: context.getString(R.string.error_generic_message)
        }
    }

    data class Database(val originalError: Throwable? = null) : AppError() {
        override val debugCode: String = "E-APP-003"
        override val message: String? = originalError?.message ?: "Database error"

        override fun getUserMessage(context: Context): String {
            return context.getString(R.string.error_generic_message)
        }
    }

    data class Unknown(val originalError: Throwable? = null) : AppError() {
        override val debugCode: String = "E-APP-004"
        override val message: String? = originalError?.message ?: "An unknown error occurred"

        override fun getUserMessage(context: Context): String {
            return context.getString(R.string.error_generic_message)
        }
    }

    /**
     * Map a throwable to an AppError
     */
    companion object {
        fun from(throwable: Throwable): AppError {
            return when (throwable) {
                is AppError -> throwable
                is IOException -> Network(throwable)
                // Add more specific mapping as needed
                else -> Unknown(throwable)
            }
        }
    }
}

File: shamelagpt-ios/shamelagpt/Core/Utilities/AppError.swift
//
//  AppError.swift
//  ShamelaGPT
//

import Foundation

/**
 * Standardized error types for the ShamelaGPT application.
 */
enum AppError: LocalizedError {
    case network(Error?)
    case auth(Error?)
    case api(Int, String?)
    case database(Error?)
    case unknown(Error?)
    
    var errorDescription: String? {
        switch self {
        case .network(let error):
            return error?.localizedDescription ?? "Network error occurred"
        case .auth(let error):
            return error?.localizedDescription ?? "Authentication error"
        case .api(let code, let message):
            return message ?? "API error (\(code))"
        case .database(let error):
            return error?.localizedDescription ?? "Database error"
        case .unknown(let error):
            return error?.localizedDescription ?? "An unknown error occurred"
        }
    }
    
    /// User message with debug code appended (for support tickets)
    var userMessageWithCode: String {
        let messageKey: String
        let debugCode: String
        
        switch self {
        case .network:
            messageKey = LocalizationKeys.noInternetConnection
            debugCode = "E-APP-001"
        case .auth:
            messageKey = LocalizationKeys.somethingWentWrong
            debugCode = "E-APP-002"
        case .api(let code, _):
            messageKey = LocalizationKeys.somethingWentWrong
            debugCode = "E-APP-\(code)"
        case .database:
            messageKey = LocalizationKeys.somethingWentWrong
            debugCode = "E-APP-003"
        case .unknown:
            messageKey = LocalizationKeys.somethingWentWrong
            debugCode = "E-APP-004"
        }
        
        return UserErrorFormatter.format(messageKey: messageKey, code: debugCode)
    }
    
    /**
     * Map a generic error to an AppError
     */
    static func mapping(_ error: Error) -> AppError {
        if let appError = error as? AppError {
            return appError
        }
        
        let nsError = error as NSError
        if nsError.domain == NSURLErrorDomain {
            return .network(error)
        }
        
        // Add more specific mapping as needed
        return .unknown(error)
    }
}
